VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,537 @@
+======================================================
+vers: a mostly universal version range specifier
+======================================================
+
+This specification is a new syntax for dependency and vulnerable version ranges.
+
+
+Context
+--------
+
+Software package version ranges and version constraints are essential:
+
+- When resolving the dependencies of a package to express which subset of the
+  versions are supported. For instance a dependency or requirement statement
+  such as "I require package foo, version 2.0 and later versions" defines a
+  range of acceptable foo versions.
+
+- When stating that a known vulnerability or bug affects a range of package
+  versions. For instance a security advisory such as "vulnerability 123 affects
+  package bar, version 3.1 and version 4.2 but not version 5" defines a range of
+  vulnerable "bar" package versions.
+
+Version ranges can be replaced by a list enumerating all the versions of
+interest. But in practice, all the versions may not yet exist when defining an
+open version range such as "v2.0 or later".
+
+Therefore, a version range is a necessary, compact and practical way to
+reference multiple versions rather than listing all the versions.
+
+
+Problem
+--------
+
+Several version range notations exist and have evolved separately to serve the
+specific needs of each package ecosystem, vulnerability databases and tools.
+
+There is no (mostly) universal notation for version ranges and there is no
+universal way to compare two versions, even though the concepts that exist in
+most version range notations are similar.
+
+Each package type or ecosystem may define their own ranges notation and version
+comparison semantics for dependencies. And for security advisories, the lack of
+portable notation for vulnerable package version ranges means that these ranges
+may be either ambiguous or hard to compute and are typically replaced by
+complete enumerations of all versions, such as the NVD CPEs.
```

Comment:
I don't think NVD CPEs provide complete enumerations.

Reply:
You are right, this is a bit more complex indeed.
1. there are version ranges as part of the configuration https://github.com/nexB/univers/pull/11/files#diff-dda3c48de6da2f63bfa5902e41755b27a2f1d8c186a21dcc2641c42d4f5cfaebR450
2. and there are daily recomputed concrete ranges available as a feed at https://nvd.nist.gov/vuln/data-feeds#cpeMatch ... which look to me as complete enumerations.

This feed contains both ranges and the enumerations. I wonder if they compute these or add them by hand?